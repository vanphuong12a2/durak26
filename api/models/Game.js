/**
 * Game.js
 *
 * @description :: A model definition represents a database table/collection.
 * @docs        :: https://sailsjs.com/docs/concepts/models-and-orm/models
 */

module.exports = {

  attributes: {

    numberOfCardsToDeal: {
      type: 'integer',
      defaultsTo: 52
    },
    trumpCard: {
      type: 'string',
      required: false
    },
    tableCards: {
      type: 'json',
      defaultsTo: []
    },
    cardsToDeal: {
      type: 'json',
      defaultsTo: []
    },
    players: {
      collection: 'player',
      via: 'gameId'
    },
  },

};
