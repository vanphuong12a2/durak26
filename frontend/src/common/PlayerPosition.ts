export enum PlayerPosition {
  BOTTOM = 'BOTTOM',
  TOP = 'TOP',
  RIGHT = 'RIGHT',
  LEFT = 'LEFT',
}

export const allPlayerPositions = [PlayerPosition.BOTTOM, PlayerPosition.RIGHT, PlayerPosition.TOP, PlayerPosition.LEFT];
