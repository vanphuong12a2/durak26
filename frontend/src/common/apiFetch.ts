const apiFetch = (url: string, params: any = {}) =>
  fetch(url, params)
    .then(response => {
      if (response.ok) {
        return response;
      } else {
        throw new Error('Status code is not 200');
      }
    })
    .then(res => res.json());

export default apiFetch;
