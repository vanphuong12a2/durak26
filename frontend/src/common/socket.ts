import socketIOClient from 'socket.io-client';
import sailsIOClient from 'sails.io.js';

const io = sailsIOClient(socketIOClient);

if (process.env.NODE_ENV !== 'production') {
  io.sails.url = 'http://localhost:1337';
}

const socket = io.socket;

export default socket;
