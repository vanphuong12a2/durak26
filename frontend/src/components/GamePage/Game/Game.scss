.game {
  display: flex;
  flex-flow: row wrap;
  padding: 5px;

  .deal-area {
    order: 1;
    flex-basis: 25%;
  }
  .player-area.top {
    order: 2;
    flex-basis: 50%;
    padding-top: 10%;
  }
  .control-area {
    order: 3;
    flex-basis: 25%;
  }
  .player-area.left {
    flex-basis: 25%;
    order: 4;
  }
  .table {
    order: 5;
    flex-basis: 50%;
  }
  .player-area.right {
    flex-basis: 25%;
    order: 6;
  }
  .player-area.bottom {
    flex-basis: 100%;
    order: 7;
  }
}
