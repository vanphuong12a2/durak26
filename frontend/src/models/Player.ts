class Player {
  public id: number;
  public name: string;
  public numberOfCards: number;


  constructor(id: number, name: string, numberOfCards: number) {
    this.id = id;
    this.name = name;
    this.numberOfCards = numberOfCards;
  }
}

export default Player;
