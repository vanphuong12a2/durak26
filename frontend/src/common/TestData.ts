import {Card} from '../components/GamePage/Game/CardSet/Card/Card';
import createMockStore, {MockStoreCreator, MockStoreEnhanced} from 'redux-mock-store';
import {ApplicationState} from '../store/ApplicationState';
import thunk, {ThunkDispatch} from 'redux-thunk';
import {AnyAction} from 'redux';
import Player from '../models/Player';

type DispatchExts = ThunkDispatch<ApplicationState, undefined, AnyAction>
const mockStoreCreator: MockStoreCreator<ApplicationState, DispatchExts> =
  createMockStore<ApplicationState, DispatchExts>([thunk]);

const getInitialApplicationState = () => (
  {
    game: {
      loading: false,
      newGame: {
        loading: false
      }
    },
    card: {
      numberOfCardsToDeal: 52,
      tableCards: [],
      playingPlayerCards: []
    },
    player: {
      players: [],
      playingPlayer: {
        loading: false
      }
    }
  }
);

export const aCard = new Card('2', 'D');
export const aPlayer = new Player('someId', 'some-name', 0);
export const freshTestStore = () => mockStoreCreator(getInitialApplicationState());
export const testStore: MockStoreEnhanced<ApplicationState, DispatchExts> = freshTestStore();
