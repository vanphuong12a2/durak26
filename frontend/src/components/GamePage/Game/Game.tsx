import React from 'react';
import ControlArea from './ControlArea/ControlArea';
import DealArea from './DealArea/DealArea';
import PlayerArea from './PlayerArea/PlayerArea';
import Player from '../../../models/Player';
import Card from '../../../models/Card';
import Table from './Table/Table';
import './Game.scss';


interface Props {
  numberOfCardsToDeal: number
  trumpCard: Card
  players: Player[]
}

const Game = (props: Props) => {

  function getPlayerAreas() {
    return new Array(4).fill(0).map((value, index) => {
      if (index < props.players.length) {
        return (
          <PlayerArea
            key={index}
            position={index}
            player={props.players[index]}
          />);
      } else {
        return (
          <PlayerArea
            key={index}
            position={index}
          />);
      }
    });
  }

  return (
    <div className='game'>
      <ControlArea/>
      <DealArea
        numberOfCardsToDeal={props.numberOfCardsToDeal}
        trumpCard={props.trumpCard}
      />
      {getPlayerAreas()}
      <Table/>
    </div>
  );
};

export default Game;
