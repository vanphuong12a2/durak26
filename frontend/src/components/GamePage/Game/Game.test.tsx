import React from 'react';
import {create} from 'react-test-renderer';
import Game from './Game';
import {aCard} from '../../../common/TestData';
import Player from '../../../models/Player';

describe('<Game />', () => {

  describe('Snapshots', () => {
    it('renders correctly', () => {
      const component = create(<Game
        numberOfCardsToDeal={2}
        trumpCard={aCard}
        players={[
          new Player(0, 'Luffy', 2),
          new Player(1, 'Sanji', 25),
          new Player(2, 'Nami', 5),
          new Player(3, 'Robin', 0)
        ]}/>
      );
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

});
