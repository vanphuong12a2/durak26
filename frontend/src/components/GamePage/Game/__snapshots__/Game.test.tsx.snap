// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> Snapshots renders correctly 1`] = `
<div
  className="game"
>
  <div
    className="control-area"
  >
    <a
      className="waves-effect waves-light btn-small"
      onClick={[Function]}
    >
      New game
    </a>
    <a
      className="waves-effect waves-light red lighten-2 btn-small"
      onClick={[Function]}
    >
      Exit
    </a>
  </div>
  <div
    className="deal-area"
  >
    <div
      className="card-set"
    >
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
    </div>
    <img
      className="card"
      src="2D.svg"
    />
  </div>
  <div
    className="player-area bottom"
  >
    <div
      className="card-set"
    >
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
    </div>
  </div>
  <div
    className="player-area top"
  >
    <div
      className="card-set"
    >
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
    </div>
  </div>
  <div
    className="player-area left"
  >
    <div
      className="card-set"
    >
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
      <img
        className="card"
        src="card_back.svg"
      />
    </div>
  </div>
  <div
    className="player-area right"
  >
    <div
      className="card-set"
    />
  </div>
  <div
    className="table"
  >
    <div
      className="table-inside"
    >
      <div
        className="table-content"
      />
    </div>
  </div>
</div>
`;
